### PR TITLE
EES-2467 chart group by filters

### PR DESCRIPTION
**Pull request by amyb-hiveit**
Adds an option to select a specific filter when grouping a chart by filter.

Changes the group by options to a radio group as that works best with conditionally showing the filters dropdown.

The filters dropdown is hidden behind the `showGroupByFilter` flag as backend work to store the value is required (EES-2620)

groupby

### CODE REVIEW

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/ChartAxisConfiguration.tsx [hunk 1]
```diff
@@ -105,14 +107,41 @@ const ChartAxisConfiguration = ({
     );
 
     if (canGroupByFilters) {
-      options.push({
-        label: 'Filters',
-        value: 'filters',
-      });
+      // EES-2467 remove this check when BE done.
+      if (showGroupByFilter) {
+        const categories: SelectOption[] = Object.keys(meta.filters)
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/ChartAxisConfiguration.tsx:112]: Could make this is a bit nicer with `Object.entries`

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/ChartAxisConfiguration.tsx [hunk 2]
```diff
@@ -105,14 +107,41 @@ const ChartAxisConfiguration = ({
...
+          .filter(filter => {
+            return meta.filters[filter].options.length > 1;
+          })
+          .map(filter => {
+            return {
+              label: filter,
+              value: meta.filters[filter].name,
+            };
+          });
+        categories.push({ label: 'All filters', value: '' });
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/ChartAxisConfiguration.tsx:122]: Should this go first?

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/ChartAxisConfiguration.tsx [hunk 3]
```diff
@@ -366,6 +387,17 @@ const ChartAxisConfiguration = ({
                   />
                 )}
 
+                {validationSchema.fields.groupBy && (
+                  <>
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/ChartAxisConfiguration.tsx:391]: Don't need the wrapping fragment

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartAxisConfiguration.test.tsx [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+      submitCount: 0,
+      id: 'options',
+      title: 'Options',
+    },
+    legend: {
+      isValid: true,
+      submitCount: 0,
+      id: 'legend',
+      title: 'Legend',
+    },
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartAxisConfiguration.test.tsx:34]: Think we can omit this form as we don't need it for testing this component?

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartAxisConfiguration.test.tsx [hunk 5]
```diff
@@ -0,0 +1,521 @@
...
+import noop from 'lodash/noop';
+import {
+  fireEvent,
+  render,
+  screen,
+  waitFor,
+  within,
+} from '@testing-library/react';
+
+describe('ChartAxisConfiguration', () => {
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartAxisConfiguration.test.tsx:21]: Nice! Gradually getting more coverage on charts :100:

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartAxisConfiguration.test.tsx [hunk 6]
```diff
@@ -0,0 +1,521 @@
...
+        unit: '',
+        label: {
+          text: '',
+        },
+      };
+      expect(handleSubmit).toHaveBeenCalledWith(formValues);
+    });
+  });
+
+  test('adding reference lines', async () => {
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartAxisConfiguration.test.tsx:387]: Might want to group these related tests into their own `describe('reference lines')` block

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartAxisConfiguration.test.tsx [hunk 7]
```diff
@@ -0,0 +1,521 @@
...
+        label: {
+          text: '',
+        },
+      };
+      expect(handleSubmit).toHaveBeenCalledWith(formValues);
+    });
+  });
+});
+
+// reference lines
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartAxisConfiguration.test.tsx:519]: We've already added these tests? Or are there more?

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartAxisConfiguration.test.tsx [hunk 8]
```diff
@@ -0,0 +1,521 @@
...
+
+// group by
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartAxisConfiguration.test.tsx:521]: Feels like we might want to at least cover some basic group filter cases as this is what the ticket is concerned with

File: src/explore-education-statistics-common/src/modules/charts/util/createDataSetCategories.ts [hunk 9]
```diff
@@ -151,8 +151,17 @@ function getCategoryFilters(
...
-        .filter(filterGroup => filterGroup.options.length > 1)
+        .filter(filterGroup => {
+          if (groupByFilter) {
+            return (
+              filterGroup.options.length > 1 &&
+              filterGroup.name === groupByFilter
+            );
+          }
+          return filterGroup.options.length > 1;
+        })
```
Comment (ntsim) [inline src/explore-education-statistics-common/src/modules/charts/util/createDataSetCategories.ts:164]: Nice! I'm surprised this ended up being so easy in the end :smile:

File: src/explore-education-statistics-common/src/modules/charts/util/__tests__/createDataSetCategories.test.ts [hunk 10]
```diff
@@ -1315,4 +1315,84 @@ describe('createDataSetCategories', () => {
...
+          filters: ['ethnicity-major-chinese', 'state-funded-secondary'],
+        },
+      ],
+      referenceLines: [],
+      visible: true,
+      unit: '',
+      min: 0,
+    };
+
+    const fullTable = mapFullTable({
```
Comment (ntsim) [inline src/explore-education-statistics-common/src/modules/charts/util/__tests__/createDataSetCategories.test.ts:1342]: So much nesting :laughing: 

Think a little improvement we could make here would be to use Immer to create a new table data with the modifications to the original.

File: src/explore-education-statistics-common/src/modules/charts/util/__tests__/createDataSetCategories.test.ts [hunk 11]
```diff
@@ -1315,4 +1315,84 @@ describe('createDataSetCategories', () => {
...
+                    label: 'State-funded secondary',
+                    value: 'state-funded-secondary',
+                  },
+                ],
+              },
+            },
+          },
+        },
+      },
+    } as TableDataResponse);
```
Comment (ntsim) [inline src/explore-education-statistics-common/src/modules/charts/util/__tests__/createDataSetCategories.test.ts:1386]: Don't think we need this type assertion?

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartAxisConfiguration.test.tsx [hunk 12]
```diff
@@ -0,0 +1,521 @@
...
+        tickConfig: 'default',
+        tickSpacing: 1,
+        type: 'major',
+        visible: true,
+        unit: '',
+        label: {
+          text: '',
+        },
+      };
+      expect(handleSubmit).toHaveBeenCalledWith(formValues);
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartAxisConfiguration.test.tsx:456]: Would probably suggest testing the submission with reference lines separately

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartAxisConfiguration.test.tsx [hunk 13]
```diff
@@ -0,0 +1,521 @@
...
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartAxisConfiguration.test.tsx:514]: Similarly, would probably suggest testing the submission without a reference line separately

**Review by ntsim — CHANGES_REQUESTED** [review]

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartAxisConfiguration.test.tsx (hunk 7 above)
Comment (amyb-hiveit) [reply src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartAxisConfiguration.test.tsx:519]: oops, notes to self left in by accident!

File: src/explore-education-statistics-common/src/modules/charts/util/__tests__/createDataSetCategories.test.ts (hunk 10 above)
Comment (amyb-hiveit) [reply src/explore-education-statistics-common/src/modules/charts/util/__tests__/createDataSetCategories.test.ts:1342]: Nice, I didn't know about that. Much better!

**Review by duncan-at-hiveit — APPROVED** [review]

**Pull request merged**